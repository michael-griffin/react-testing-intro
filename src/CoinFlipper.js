import React, { useState } from "react";

const HEADS_URL =
  "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/2017" +
  "-D_Roosevelt_dime_obverse_transparent.png/912px-2017-D_Roosev" +
  "elt_dime_obverse_transparent.png";
const TAILS_URL =
  "https://upload.wikimedia.org/wikipedia/commons/thumb/d/d9/2017" +
  "-D_Roosevelt_dime_reverse_transparent.png/911px-2017-D_Roosev" +
  "elt_dime_reverse_transparent.png";

function randomFlip() {
  return Math.random() > 0.5 ? "heads" : "tails";
}

function CoinFlipper() {
  //have some state for our coin
  //const [isHeads, setIsHeads ] = useState();
  //have some state for how many times the coin is flipped
  const [flips, setFlips] = useState([]);

  function handleClick() {
    const nextFlip = randomFlip();
    setFlips(() => {
      return [...flips, nextFlip];
    });
  }

  let imgSource;
  if (flips.length > 0) {
    imgSource = flips[flips.length - 1] === "heads" ? HEADS_URL : TAILS_URL;
  }
  const headsCount = flips.filter((f) => f === "heads").length;

  return (
    <div>
      <h1>Let's flip a coin!</h1>
      <img src={imgSource} style={{ width: "200px" }} className="flip-image" />
      <br></br>
      <button onClick={handleClick} className="flipButton">
        Flip Me!
      </button>
      <p>
        Out of {flips.length}, there have been {headsCount} heads and{" "}
        {flips.length - headsCount} tails.
      </p>
    </div>
  );
}

export default CoinFlipper;

export { randomFlip };
