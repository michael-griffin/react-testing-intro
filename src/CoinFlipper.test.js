import CoinFlipper from "./CoinFlipper";
import * as cf from "./CoinFlipper";
cf.randomFlip = jest.fn();

import { render, fireEvent } from "@testing-library/react";

it("renders without breaking", function () {
  render(<CoinFlipper />);
});

it("snapshot test - renders the same", function () {
  const { container } = render(<CoinFlipper />);

  expect(container).toMatchSnapshot();
});

it("image url changes with flip", function () {
  cf.randomFlip.mockReturnValueOnce("heads");

  const { container } = render(<CoinFlipper />);
  fireEvent.click(container.querySelector(".flipButton"));

  expect(container.querySelector(".flip-image").getAttribute("src")).toContain(
    "obverse"
  );
});
