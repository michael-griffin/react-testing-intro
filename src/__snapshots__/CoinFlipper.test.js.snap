// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test - renders the same 1`] = `
<div>
  <div>
    <h1>
      Let's flip a coin!
    </h1>
    <img
      class="flip-image"
      style="width: 200px;"
    />
    <br />
    <button
      class="flipButton"
    >
      Flip Me!
    </button>
    <p>
      Out of 
      0
      , there have been 
      0
       heads and
       
      0
       tails.
    </p>
  </div>
</div>
`;
